rllib/env/single_agent_episode.py

```diff
@@ -295,159 +283,187 @@ def is_done(self) -> bool:
-        """Converts a `SingleAgentEpisode` into a `SampleBatch`.
+    def get_data_dict(self):
+        """Converts a `SingleAgentEpisode` into a data dict mapping str keys to data.
 
-        Note that `RLlib` is relying in training on the `SampleBatch`  class and
-        therefore episodes have to be converted to this format before training can
-        start.
+        The keys used are:
+        SampleBatch.EPS_ID, T, OBS, INFOS, ACTIONS, REWARDS, TERMINATEDS, TRUNCATEDS,
+        and those in `self.extra_model_outputs`.
 
         Returns:
-            An `ray.rLlib.policy.sample_batch.SampleBatch` instance containing this
-            episode's data.
+            A data dict mapping str keys to data records.
         """
-        return SampleBatch(
+        t = list(range(self.t_started, self.t))
+        terminateds = [False] * (len(self) - 1) + [self.is_terminated]
+        truncateds = [False] * (len(self) - 1) + [self.is_truncated]
+        eps_id = [self.id_] * len(self)
+
+        if self.is_numpy:
+            t = np.array(t)
+            terminateds = np.array(terminateds)
+            truncateds = np.array(truncateds)
+            eps_id = np.array(eps_id)
+
+        return dict(
             {
-                SampleBatch.EPS_ID: np.array([self.id_] * len(self)),
-                SampleBatch.OBS: self.observations[:-1],
-                SampleBatch.NEXT_OBS: self.observations[1:],
-                SampleBatch.ACTIONS: self.actions,
-                SampleBatch.REWARDS: self.rewards,
-                SampleBatch.T: list(range(self.t_started, self.t)),
-                SampleBatch.TERMINATEDS: np.array(
-                    [False] * (len(self) - 1) + [self.is_terminated]
-                ),
-                SampleBatch.TRUNCATEDS: np.array(
-                    [False] * (len(self) - 1) + [self.is_truncated]
-                ),
-                # Return the infos after stepping the environment.
-                SampleBatch.INFOS: self.infos[1:],
-                **self.extra_model_outputs,
-            }
+                # Trivial 1D data (compiled above).
+                SampleBatch.TERMINATEDS: terminateds,
+                SampleBatch.TRUNCATEDS: truncateds,
+                SampleBatch.T: t,
+                SampleBatch.EPS_ID: eps_id,
+                # Retrieve obs, infos, actions, rewards using our get_... APIs,
+                # which return all relevant timesteps (excluding the lookback
+                # buffer!).
+                SampleBatch.OBS: self.get_observations(),
+                SampleBatch.INFOS: self.get_infos(),
+                SampleBatch.ACTIONS: self.get_actions(),
+                SampleBatch.REWARDS: self.get_rewards(),
+            },
+            # All `extra_model_outs`: Same as obs: Use get_... API.
+            **{
+                k: self.get_extra_model_outputs(k)
+                for k in self.extra_model_outputs.keys()
+            },
         )
 
-    @staticmethod
```

Comment:
We'll try to get rid of SampleBatch. There is no application currently that requires constructing an episode from an existing SampleBatch (only the other way around: Episode -> SampleBatch)